Nginx custom crds

sarthyparty opened the pull request:
Can I have some feedback on these decisions I made:
https://docs.google.com/document/d/1NTZskx2lNcrBh9-VPLh-p_MIl35NeFIDpX4XJwynHPU/edit?tab=t.0

Commit:
Add NGINX custom CRDs conversion

sjberman commented on pkg/i2gw/providers/nginx/annotations/header_manipulation.go:
```diff
@@ -21,6 +21,8 @@ import (
 	"github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw/intermediate"
 	"github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw/providers/common"
+	ncommon "github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw/providers/nginx/common"
+	fcommon "github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw/providers/nginx/common/filters"
```
I'd just keep this import name as filters.

sjberman commented on pkg/i2gw/providers/nginx/common/policies/factory.go:
```diff
@@ -0,0 +1,229 @@
+const (
+	BackendTLSPolicyKind = "BackendTLSPolicy"
+	GRPCRouteKind        = "GRPCRoute"
+	ServiceKind          = "Service"
```
These aren't policies, so let's we don't use that terminology.

sjberman commented on pkg/i2gw/providers/nginx/common/types.go:
```diff
@@ -0,0 +1,92 @@
+// GetNotifications returns empty slice since notifications are dispatched immediately
+func (c *DispatchNotificationCollector) GetNotifications() []notifications.Notification {
+	return []notifications.Notification{}
+}
```
Be sure your files have newlines at the end.

sjberman commented on pkg/i2gw/providers/nginx/crds/vs_gateway.go:
```diff
@@ -0,0 +1,338 @@
+		Name:      gatewayName,
+	}
+
+	// Collect all hostnames and TLS configurations
```
Doesn't look like we're getting TLS config here, just hostnames.

sarthyparty replied:
I refactored this code

sjberman commented on pkg/i2gw/providers/nginx/crds/vs_gateway.go:
```diff
@@ -0,0 +1,338 @@
+				Port:     httpPort,
+				Protocol: gatewayv1.HTTPProtocolType,
+				Hostname: host,
+				Secret:   "", // HTTP has no secret
```
No need to specify this, it'll default to empty string.

sjberman commented on pkg/i2gw/providers/nginx/fixtures/virtualserver/output/nginx-vs-basic-routing.yaml:
```diff
@@ -0,0 +1,101 @@
+    protocol: HTTPS
+    tls:
+      certificateRefs:
+      - group: null
+        kind: null
```
These should be set

sarthyparty replied:
The kind should be Secret and group empty string for core?

sjberman replied:
Yes, sorry should have been more specific.

sjberman commented on pkg/i2gw/providers/nginx/fixtures/virtualserver/output/nginx-vs-basic-routing.yaml:
```diff
@@ -0,0 +1,101 @@
+  - filters:
+    - requestRedirect:
+        path:
+          replaceFullPath: https://docs.example.com
```
This isn't going to work as expected. If it's a URL we need to redirect to, we need to utilize the `hostname` field (and `scheme` if changed). https://gateway-api.sigs.k8s.io/reference/spec/#httprequestredirectfilter

sarthyparty replied:
Oh I see

sjberman commented on pkg/i2gw/providers/nginx/fixtures/virtualserver/output/nginx-vs-advanced-proxy.yaml:
```diff
@@ -0,0 +1,127 @@
+  - backendRefs:
+    - name: backend-v1-service
+      port: 8080
+    filters:
```
Missing response headers from this one.

sarthyparty replied:
Oh thanks for the catch

sarthyparty replied:
it was an issue in the input virtualserverSpec, I was using set instead of add

sjberman commented on pkg/i2gw/providers/nginx/fixtures/virtualserver/output/nginx-vs-custom-listeners.yaml:
```diff
@@ -0,0 +1,67 @@
+  namespace: default
+spec:
+  gatewayClassName: nginx
+  listeners:
+  - hostname: custom.example.com
+    name: http-8080-custom-example-com
+    port: 8080
+    protocol: HTTP
```
Looks like we're missing the 8443 listener?

sarthyparty replied:
good catch thanks

sarthyparty replied:
it was an issue on my global configuration setup

sjberman commented on pkg/i2gw/providers/nginx/fixtures/virtualserver/output/nginx-vs-grpc.yaml:
```diff
@@ -0,0 +1,132 @@
+status: {}
+---
+apiVersion: gateway.networking.k8s.io/v1
+kind: HTTPRoute
```
Wrong route type

sarthyparty replied:
This one was funny I actually just forgot to mark the upstream as GRPC in the virtualserverSpec lol

sjberman commented on pkg/i2gw/providers/nginx/fixtures/virtualserver/output/nginx-vs-grpc.yaml:
```diff
@@ -0,0 +1,132 @@
+  - backendRefs:
+    - name: http-service
+      port: 8080
+    matches:
+    - path:
+        type: PathPrefix
+        value: /api/rest
```
Wrong format for GRPCRoute matches. See https://github.com/nginx/nginx-gateway-fabric/blob/main/examples/grpc-routing/exact-method.yaml as an example.

sjberman commented on pkg/i2gw/providers/nginx/fixtures/virtualserver/output/nginx-vs-grpc.yaml:
```diff
@@ -0,0 +1,132 @@
+  - matches:
+    - path:
+        type: PathPrefix
+        value: /api.DataService
```
Weighted backends aren't showing up in here

sjberman commented on pkg/i2gw/providers/nginx/fixtures/virtualserver/output/nginx-vs-traffic-splitting.yaml:
```diff
@@ -0,0 +1,99 @@
+        type: PathPrefix
+        value: /redirect-test
+status:
+  parents: null
```
There's no traffic splitting defined in this Route

sjberman commented on pkg/i2gw/providers/nginx/fixtures/virtualserver/output/nginx-vs-tls-redirect.yaml:
```diff
@@ -0,0 +1,80 @@
+        type: PathPrefix
+        value: /api
+status:
+  parents: null
```
Not seeing a redirect filter in here.

sarthyparty replied:
should be fixed now.

sarthyparty commented:
@sjberman I am handling the case of virtualservers with the same port, protocol (https), hostname, but different secret, because Gateway API will reject two listeners and NGF will reject multiple tls configs. Currently I'm creating an entirely new gateway for this but I'd much rather add another tls certificate ref because it seems like the correct and future proof solution. Do you think NGF will support multiple TLS certs soon and I can potentially use that solution?

bjee19 commented on pkg/i2gw/intermediate/provider_nginx.go:
```diff
@@ -16,16 +16,136 @@ limitations under the License.
+	StatusCode *int
+}
+
+type NginxLoadBalancingConfig struct {
+	Method string // round_robin, least_conn, ip_hash, etc.
+}
```
nit:

For this, usually something that follows this structure would be better:

```
type LoadBalacingMethod string

const (

roundRobin LoadBalancingMethod = "round_robin"

leastConn LoadBalancingMethod = "least_conn"

...

type NginxLoadBalancingConfig struct {
    Method LoadBalacingMethod
}

```

bjee19 replied:
And that applies to all the other fields here that are enums

sarthyparty replied:
I'm removing this stuff because to me there seems to be no point in storing this information

bjee19 commented on pkg/i2gw/providers/nginx/annotations/header_manipulation_helpers_test.go:
i don't think a separate test file for the helpers is needed, i think you could combine them

bjee19 commented on pkg/i2gw/providers/nginx/common/filters/factory.go:
```diff
@@ -0,0 +1,340 @@
+
+// Header defines an HTTP Header.
+type Header struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
```
how come some of these have json values and some don't?

sarthyparty replied:
The json values aren't needed so I removed them

bjee19 commented on pkg/i2gw/providers/nginx/common/filters/factory.go:
```diff
@@ -0,0 +1,340 @@
+// RequestRedirectOptions contains options for request redirect filters
+type RequestRedirectOptions struct {
+	// HTTP status code for the redirect
+	StatusCode int
```
nit: for style, the comment should start with the name of the variable

```suggestion
	// StatusCode represents the HTTP status code for the redirect
	StatusCode int
```

bjee19 commented on pkg/i2gw/providers/nginx/common/filters/factory.go:
```diff
@@ -0,0 +1,340 @@
+				})
+			} else {
+				skippedHeaders++
+			}
```
This is similar to the logic above, but it is reversed, can you make them match?

bjee19 commented on pkg/i2gw/providers/nginx/common/filters/factory.go:
```diff
@@ -0,0 +1,340 @@
+				fmt.Sprintf("Request header modifications converted (%d headers)", len(headers)),
+				sourceObj)
+		}
+		if notificationCollector != nil && skippedHeaders > 0 {
+			notificationCollector.AddWarning(
+				fmt.Sprintf("%d request headers with NGINX variables skipped - Gateway API doesn't support dynamic values", skippedHeaders),
+				sourceObj)
+		}
```
if you want to have this included, it should be included in the above if check also, or else skippedHeaders is not used in the above if check

bjee19 replied:
i guess just pull it out of both the if checks

bjee19 commented on pkg/i2gw/providers/nginx/common/filters/factory.go:
```diff
@@ -0,0 +1,340 @@
+}
+
+// createResponseHeaderModifierFilter creates a ResponseHeaderModifier filter
+func createResponseHeaderModifierFilter(headerOpts *HeaderModifierOptions, notificationCollector common.NotificationCollector, sourceObj client.Object) *gatewayv1.HTTPRouteFilter {
```
i feel like this function and the request modifier function can have a lot of the code be refactored into a generic function

bjee19 commented on pkg/i2gw/providers/nginx/common/resources/factory.go:
```diff
@@ -0,0 +1,229 @@
+	// Source label for tracking the origin (e.g., "nginx-ssl-services")
+	SourceLabel string
+	// Additional labels to apply
+	Labels map[string]string
```
same comment as above with starting the comment with the name of the variable

bjee19 commented on pkg/i2gw/providers/nginx/common/resources/factory.go:
```diff
@@ -0,0 +1,229 @@
+					},
+				},
+			},
+			Validation: gatewayv1alpha3.BackendTLSPolicyValidation{
+				// Note: WellKnownCACertificates and Hostname fields are intentionally left empty
+				// These fields must be manually configured based on your backend service's TLS setup
+			},
```
is this ai generated part needed?

sarthyparty replied:
brooo I'm pretty sure I wrote that intentionally. https://github.com/kubernetes-sigs/ingress2gateway/pull/224#discussion_r2195537495

bjee19 replied:
oh i see, that comment makes sense, though this message is just a comment isn't it? does that mean these fields will need to be configured in the code itself or can a user see this message when running the tool?

sarthyparty replied:
yeah there is a notification being added as well

bjee19 commented on pkg/i2gw/providers/nginx/common/resources/factory.go:
```diff
@@ -0,0 +1,229 @@
+	grpcOpts := opts.GRPCRoute
+
+	// Build labels
+	labels := map[string]string{
+		"app.kubernetes.io/managed-by": "ingress2gateway",
+	}
+	if grpcOpts.SourceLabel != "" {
+		labels["ingress2gateway.io/source"] = grpcOpts.SourceLabel
+	}
+	for k, v := range grpcOpts.Labels {
+		labels[k] = v
+	}
```
can refactor this duplicate code out and create a separate function

sarthyparty replied:
I'm removing it because I'm not gonna use it for now

bjee19 commented on pkg/i2gw/providers/nginx/crds/condition_matcher.go:
```diff
@@ -0,0 +1,178 @@
+	}
+}
+
+// containsRegexPatterns is implemented in utils.go
```
this doesn't need to be here

bjee19 commented on pkg/i2gw/providers/nginx/crds/condition_matcher.go:
lots of duplicate code in the functions here that should be refactored out into separate functions.

sarthyparty replied:
refactored it now

bjee19 commented on pkg/i2gw/providers/nginx/crds/conversion_main.go:
```diff
@@ -0,0 +1,305 @@
+
+// CRDsToGatewayIR converts nginx VirtualServer, VirtualServerRoute, and TransportServer CRDs to Gateway API resources
+// This function creates one shared Gateway per namespace that handles both Layer 7 and Layer 4 traffic
+func CRDsToGatewayIR(
```
nit: probably better name would be ConvertCRDsToGatewayIR

bjee19 commented on pkg/i2gw/providers/nginx/crds/conversion_main.go:
```diff
@@ -0,0 +1,305 @@
+			}
+		}
+		addNotification(&notificationList, notifications.InfoNotification,
+			"GlobalConfiguration found", nil)
```
i feel like this isn't needed

bjee19 commented on pkg/i2gw/providers/nginx/crds/conversion_main.go:
```diff
@@ -0,0 +1,305 @@
+			}
+			for routeKey, route := range udpRoutes {
+				udpRouteMap[routeKey] = route
+			}
```
can we use https://pkg.go.dev/maps#Copy

bjee19 commented on pkg/i2gw/providers/nginx/crds/route_converter.go:
```diff
@@ -0,0 +1,658 @@
+	return specificMatch
+}
+
+// handleRouteActions processes different route action types
+func (c *VirtualServerRouteConverter) handleActions(vs nginxv1.VirtualServer, action *nginxv1.Action, rule *gatewayv1.HTTPRouteRule) {
```
```suggestion
// handleRouteActions processes different route action types
func (c *VirtualServerRouteConverter) handleRouteActions(vs nginxv1.VirtualServer, action *nginxv1.Action, rule *gatewayv1.HTTPRouteRule) {
```

bjee19 commented on pkg/i2gw/providers/nginx/fixtures/transportserver/input/nginx-ts-no-listener.yaml:
what does no listener mean? doesn't the transportServer have a listener?  and the comment says 

> Mixed VirtualServer and TransportServer in same namespace

sarthyparty replied:
Oh I made a test for a case where ts didn't reference a listener and wanted to make sure my code handled it right, I'll remove it

bjee19 commented on pkg/i2gw/providers/nginx/crds/upstream_converter.go:
```diff
@@ -0,0 +1,301 @@
+)
+
+// UpstreamConfig represents upstream configuration for conversion
+type UpstreamConfig struct {
+	Name                string                 // The name of the upstream. Must be a valid DNS label as defined in RFC 1035.
+	Service             string                 // The name of a service. The service must belong to the same namespace as the resource.
+	Subselector         map[string]string      // Selects the pods within the service using label keys and values.
+	UseClusterIP        *bool                  // Enables using the Cluster IP and port of the service instead of the default behavior.
+	Port                uint16                 // The port of the service. Must fall into the range 1..65535.
+	LBMethod            string                 // The load balancing method. Default is specified in the lb-method ConfigMap key.
+	FailTimeout         string                 // The time during which unsuccessful attempts should happen to consider server unavailable.
+	MaxFails            *int                   // The number of unsuccessful attempts to consider server unavailable.
+	MaxConns            *int                   // The maximum number of simultaneous active connections to an upstream server.
+	Keepalive           *int                   // Configures the cache for connections to upstream servers.
+	ConnectTimeout      string                 // The timeout for establishing a connection with an upstream server.
+	ReadTimeout         string                 // The timeout for reading a response from an upstream server.
+	SendTimeout         string                 // The timeout for transmitting a request to an upstream server.
+	NextUpstream        string                 // Specifies in which cases a request should be passed to the next upstream server.
+	NextUpstreamTimeout string                 // The time during which a request can be passed to the next upstream server.
+	NextUpstreamTries   *int32                 // The number of possible tries for passing a request to the next upstream server.
+	ClientMaxBodySize   string                 // Sets the maximum allowed size of the client request body.
+	TLS                 *nginxv1.UpstreamTLS   // The TLS configuration for the Upstream.
+	HealthCheck         *nginxv1.HealthCheck   // The health check configuration for the Upstream.
+	SlowStart           string                 // The slow start allows an upstream server to gradually recover its weight.
+	Queue               *nginxv1.UpstreamQueue // Configures a queue for an upstream.
+	Buffering           *bool                  // Enables buffering of responses from the upstream server.
+	BufferSize          string                 // Sets the size of the buffer used for reading the first part of a response.
+	NTLM                bool                   // Allows proxying requests with NTLM Authentication.
+	Type                string                 // The type of the upstream. Supported values are http and grpc.
+	Backup              string                 // The name of the backup service of type ExternalName.
+	BackupPort          *uint16                // The port of the backup service.
+	Weight              *int32                 // Weight for load balancing (used in splits)
+	SessionCookie       *nginxv1.SessionCookie // Configures session persistence for the upstream.
+}
```
I feel like this is too much, and you should group similar configs into smaller types.

bjee19 replied:
NextUpstreams, Buffer, Backup, ...

sarthyparty replied:
yes, this is actually just all the fields of an upstream in vs spec. This is probably something to bring up later, of how to handle all these fields that I don't support. I can put them into a nginx specific IR, basically a holder so that they're stored and then can be converted in the future technically

sarthyparty replied:
is it really worth grouping them, because I don't exactly need to use them its just for reading and then warning the user if they've set it that it won't be converted

bjee19 replied:
Yea if you're going to keep the warning for each field then its fine to not group them and this can be resolved.

bjee19 reviewed (commented):
good work, did a quick go through and left my feedback

rikatz committed:
Do not print the table on stdout (#233)

adrianmoisey committed:
Add parameter to ingress-nginx provider to select ingress class (#231)

* Add parameter to ingress-nginx provider to select ingress class

This allows users to select the ingress-nginx class if they use
something other than the default 'nginx'

* Add test for ingress-nginx ingress class flag

* Clarify what the test expectations are

* Make the test compare the class name

sarthyparty committed:
Add NGINX provider support (#224)

* Add NGINX provider with annotation support

* add more test

* fix tls for custom ports

* update readme

* go mod tidy

* align table

* remove kubernetes from name

* link issue in annotations readme

* move issue link

* add notifications for grpcfilter conversion

* update grpc services test

* update error message to be clearer

* fix comments on header and rewrite

* rename test functions

* switch t error instead of t fatal

* clean tests

* rework tests to use error instead of fatal

* resolve comments

* Update banner to 2025

* Update readme, remove managed by labels

* set target ref for btp to be core

* update toir and merge irs

* move grpc and btp logic to common

* fix verify

* remove core group var

* Update comment and separate btp name gen function

* remove unecessary labels

* move btp name gen function to nginx provider

* fix gofumpt

LiorLieberman committed:
Update ingress2gateway reviewers (#237)
